### PR TITLE
Fix Ghidra Import

### LINKED ISSUE #2: Fix the Ghidra import
- [ ] Wait for #1 to be merged
- [ ] Make the Ghidra import work again - currently, it fails to find the correct `.venv`
- [ ] Think about a good setup. One idea:
  - Add `isle/.venv/Lib/ghidra_import` to the list of script directories in Ghidra after `pip install`
  - Make the script search for both a `reccmp-project.yml` and a `.venv` subfolder on its path up

### PR DESCRIPTION
**Pull request by jonschz**
This change restores the ability to import a built project into Ghidra.

**Known issues and limitations**
- Currently, the `reccmp-build.yml` must be at an auto-detectable location from the Ghidra script, i.e. it should be located next to `reccmp-project.yml`. This aspect is relevant to the discussion at #7 .
- There is still one hard-coded reference to Lego Island directly, and I am not sure what the best solution is. I'd also be fine with merging this PR and opening a follow-up issue for this.

It'd be great if you could try to set it up based on the reccmp README and see if/how the explanations could be improved.

Closes #2

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -17,7 +17,7 @@ install_requires = file:requirements.txt
 
 [options.packages.find]
 exclude =
-    reccmp.tests.*
+    tests*
```
Comment (jonschz) [inline setup.cfg:20]: Drive-by fix: The exclusion of tests in `pip install` did not work correctly

File: reccmp/ghidra_scripts/lego_util/ghidra_helper.py [hunk 2]
```diff
@@ -109,7 +109,9 @@ def sanitize_name(name: str) -> str:
     # Importing function names like `FUN_10001234` into BETA10 can be confusing
     # because Ghidra's auto-generated functions look exactly the same.
     # Therefore, such function names are replaced by `LEGO_10001234` in the BETA10 import.
-    if GLOBALS.module == SupportedModules.BETA10:
+
+    # FIXME: The identification here is a crutch - we need a more reusable solution for this scenario
```
Comment (jonschz) [inline reccmp/ghidra_scripts/lego_util/ghidra_helper.py:113]: This is the one place that is still isle-specific. Please tell me if you want me to find a solution before merging.

One idea I had was to add a `ghidra` section to the targets of a project configuration like so:
```yaml
  BETA10:
    filename: BETA10.DLL
    [...]
    ghidra:
      function_prefix: "LEGO1"
```

**Review by jonschz — COMMENTED** [review]
I also chose to relocate the `ghidra_scripts` directory into the `reccmp` project because I didn't see much of a point in installing two top-level packages into the virtualenv. If you have strong opinions one way or the other, please let me know.

File: reccmp/project/detect.py [hunk 3]
```diff
@@ -118,9 +118,9 @@ def from_directory(cls, directory: Path) -> typing.Optional["RecCmpProject"]:
 class RecCmpBuiltProject:
     def __init__(
         self,
-        project_config_path: typing.Optional[Path],
-        user_config: typing.Optional[Path],
-        build_config: typing.Optional[Path],
+        project_config_path: Path,
```
[the remark this answers is not in the document]
Comment (madebr) [reply reccmp/project/detect.py:121]: This is an oversight on my part.
I started with making these members optional, but then you'd need to check for validity every time you use them.
An object of which every member can be None is of nu use.

File: reccmp/ghidra_scripts/lego_util/ghidra_helper.py (hunk 2 above)
Comment (madebr) [reply reccmp/ghidra_scripts/lego_util/ghidra_helper.py:113]: Feel free to add whatever attributes you think are needed.
It's also be useful for a "ghidra" type in https://github.com/isledecomp/reccmp/issues/8

File: reccmp/isledecomp/compare/core.py [hunk 4]
```diff
@@ -67,12 +68,16 @@ def lookup(addr: int, size: int) -> Optional[bytes]:
 class Compare:
     # pylint: disable=too-many-instance-attributes
     def __init__(
-        self, orig_bin: IsleBin, recomp_bin: IsleBin, pdb_file: str, code_dir: str
+        self,
+        orig_bin: IsleBin,
+        recomp_bin: IsleBin,
+        pdb_file: Path | str,
```
[the remark this answers is not in the document]
Comment (madebr) [reply reccmp/isledecomp/compare/core.py:74]: Wouldn't making them `Path` throughout the code make more sense? It's a form of intent and self-documenting code.

[the remark this answers is not in the document]
Comment (disinvite) [reply reccmp/isledecomp/compare/core.py:74]: `argparse` must support creating a `Path` object from the command line, right? If so, then it should be `Path` all the way down. But we could leave `Path | str` for now since this is not the only lingering type issue.

[the remark this answers is not in the document]
Comment (madebr) [reply reccmp/isledecomp/compare/core.py:74]: Yes, simply add `type=Path`
Keeping `Path | str` is fine with me :)

File: reccmp/ghidra_scripts/README.md [hunk 5]
```diff
@@ -4,18 +4,28 @@ The scripts in this directory provide additional functionality in Ghidra, e.g. i
...
 Since these scripts and its dependencies are written in Python 3, [Ghidrathon](https://github.com/mandiant/Ghidrathon) must be installed first. Follow the instructions and install a recent build (these scripts were tested with Python 3.12 and Ghidrathon v4.0.0).
 
 ### Script Directory
+This step differs slightly depending on your setup.
+
 - In Ghidra, _Open Window -> Script Manager_.
 - Click the _Manage Script Directories_ button on the top right.
-- Click the _Add_ (Plus icon) button and select this file's parent directory.
+- Click the _Add_ (Plus icon) button.
+    - Case 1: Select `Lib/site-packages/reccmp/ghidra_scripts` inside your _project's_ virtual environment.
```
Comment (disinvite) [inline reccmp/ghidra_scripts/README.md:24]: Maybe a note here that you have to enable "Show '.' files" under "File Chooser Options" if your environment is under ".venv"

**Comment by disinvite** [comment]
This one tripped me up: I had a hand-edited reccmp-build.yml under the isle dir from before we cut over to the reccmp module. If the paths to the recomp and pdb are relative (i.e. `build/LEGO1.PDB`) then the import script can't locate them.

File: reccmp/ghidra_scripts/lego_util/ghidra_helper.py (hunk 2 above)
Comment (disinvite) [reply reccmp/ghidra_scripts/lego_util/ghidra_helper.py:113]: What if we don't overwrite the Ghidra function name if the one from `reccmp` has the `FUN_` prefix? (Or if it matches a `temp_function_prefix` key from the yml?) We can still populate all the other information.

I don't know how common it is for a decomp to have two different versions of the same binary, but this might be an unusual situation.

**Comment by disinvite** [comment]
I'll keep looking at this tomorrow, but my imports are missing functions for both LEGO1 and ISLE. LEGO1 misses the very first function `Score::Score` and ISLE doesn't annotate much of anything. I'm sure it must be something I have misconfigured, but I'm not sure what.
[logdump-isle.txt](https://github.com/user-attachments/files/17536182/logdump-isle.txt)

File: reccmp/ghidra_scripts/lego_util/ghidra_helper.py (hunk 2 above)
Comment (jonschz) [reply reccmp/ghidra_scripts/lego_util/ghidra_helper.py:113]: When I was working on BETA10 I found it hard to tell if a function was not matched and had Ghidra's default name, or if it was matched and was named `FUN_...` by the Ghidra script. There are also some address collisions between LEGO1  and BETA10 functions.

The "clean" solution would be to use a different pattern in the source code for unnamed functions, but that would be a major change to the existing code. It would have one additional minor benefit: It'll be easier to tell apart non-stubbed functions from stubbed / implemented ones in LEGO1.

**Comment by jonschz** [comment]
@disinvite could you also provide a LEGO1 log? I haven't tried ISLE.EXE yet since I didn't realise that it should be possible now.

**Comment by disinvite** [comment]
Here's the log from LEGO1 after deleting items from those two header files. (I have one from last night too if you want it.) I didn't know which things were known issues; maybe `Score::Score` was never there and I didn't notice.
[legolog.txt](https://github.com/user-attachments/files/17550421/legolog.txt)

But certainly the script is working again now in the new module, and since that was the goal of this PR then I think we are good.

**Comment by jonschz** [comment]
Thanks! This is a good reminder to document the known issues in the Ghidra import, even if there is no need to fix them any time soon. I will still look into the relative import issue you discovered, and there is one more pre-existing bug I'd like to understand before merging.

File: reccmp/ghidra_scripts/lego_util/ghidra_helper.py (hunk 2 above)
Comment (jonschz) [reply reccmp/ghidra_scripts/lego_util/ghidra_helper.py:131]: will be addressed in a follow-up, see #16

**Comment by jonschz** [comment]
Ticket for known issues: #17 
Ticket for remaining isle references: #16 

>  If the paths to the recomp and pdb are relative (i.e. build/LEGO1.PDB) then the import script can't locate them.
Should be fixed now for the `reccmp-build.yml`. I haven't looked at `reccmp-user.yml` yet.

**Comment by jonschz** [comment]
@disinvite could you take a final look at the most recent changes, and verify that the issues you found are fixed? Thanks!

**Comment by disinvite** [comment]
Looks good to me!
